src/coreclr/jit/lowerxarch.cpp

```diff
@@ -3943,6 +3943,50 @@ GenTree* Lowering::TryLowerAndOpToAndNot(GenTreeOp* andNode)
     return andnNode;
 }
 
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerBswapOp: Tries to contain GT_BSWAP node when possible
+//
+// Arguments:
+//    node - GT_BSWAP node to contain
+//
+// Notes:
+//    Containment is not performed when optimizations are disabled
+//    or when MOVBE instruction set is not found
+//
+void Lowering::LowerBswapOp(GenTreeOp* node)
+{
+    assert(node->OperIs(GT_BSWAP, GT_BSWAP16));
+
+    if (!comp->opts.OptimizationEnabled() || !comp->compOpportunisticallyDependsOn(InstructionSet_MOVBE))
+    {
+        return;
+    }
+
+    GenTree* operand = node->gtGetOp1();
+    if (IsContainableMemoryOp(operand) && IsSafeToContainMem(node, operand))
+    {
+        MakeSrcContained(node, operand);
```

Comment:
What happens here if the indirection is a small indirection? E.g. we might have a `TYP_BYTE` indirection feeding into a `TYP_INT` `bswap`. I don't see anything that rejects the optimization.

Reply:
We don't generate `GT_BSWAP` for `TYP_BYTE`.

https://github.com/dotnet/runtime/blob/dc030a0362e3177389f4d4e73e1332c345eab891/src/coreclr/jit/importer.cpp#L4760

Reply:
Obviously not, that would be nonsensical :-)
I am talking about a case like:
```csharp
sbyte[] arr = {1, 2, 3, 4, 5, 6, 7};
int test = BinaryPrimitives.ReverseEndianess((int)arr[0]);
```

In fact, thinking about this a little bit more, I don't see how we can do this optimization for sign-extended 16-bit values at all. The semantics of `BSWAP16(TYP_SHORT IND(x))` is to load 16 bits from x, sign extend it to 32 bits, and then swap the lower 2 bytes. The semantics of this transformation can never be this -- at best we can efficiently load 16 bits, swap them, and then sign extend the result.

Reply:
Ah, yes. That case seems to actually miss sign extension. For 16-bit values I have already verified that we correctly sign extend. The `movbe` correctly loads the word to the lower bits of the register and swaps the bytes. After that we can sign extend the register which produces the correct result. The cast node added in the importer is enough to deal with that.

Reply:
> That case seems to actually miss sign extension. 
I don't think the problem here is sign extension. It will emit an incorrectly sized `movbe`.

> For 16-bit values I have already verified that we correctly sign extend. The movbe correctly loads the word to the lower bits of the register and swaps the bytes.

The 'problem' is that currently, the result of `TYP_SHORT IND` followed by `bswap16` has bit 7 copied in the upper 16 bits. We seemed to be relying on that but in fact it looks more like a bug. I have opened #67723 for it.

The semantics the transformation implicitly assumes here is that `bswap16` swaps the lower two bytes and leaves the upper two bytes undefined. That sounds fine to me, but clearly not to the rest of the JIT (#67723 showing why). :-)

Reply:
> I don't think the problem here is sign extension. It will emit an incorrectly sized movbe.

Yeah, thats the case. Got a bit lost in the words there, heh.

> That sounds fine to me, but clearly not to the rest of the JIT

Interesting, so we actually have had this type of bug long time..